### PR TITLE
docs: Add float16 precision warning for optimizer eps and small values

### LINKED ISSUE #20619: 16 bit precision in Trainer leading to NaN
### Description & Motivation

In Trainer, when using lightning_precision other than 32-true, small float values (~ > 1e-7 - 1e-8 ?) can be round up to zero (in input batches, model parameters and its gradients, optimizer hyper-parameters) leading to NaN. 

This does not raise any warning or error during training as the training loop ignore NaN. 
After training, in inference mode, all output can be NaN, (which makes it a poor model).

After debugging for several days, I found that I had issue in my parameter weights (both from blank initialization or from pre-trained) but also in the very commonly used Adam optimizer (with default eps=1e-8 it was updating all my parameters to NaN in the very first step, even when learning rate was set to 0).

### Pitch

I would suggest to add a caveat in the documentation to warn the user that one has to deal with every possible small values in its model, optimizer and input data.
Also some default warnings when autocasts to lower-precision lead to over/under -flow would be nice. 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bhimrazy**
### What does this PR do?

Adds a warning to the 16-bit precision documentation about float16's inability to represent values smaller than ~6e-5, which can cause NaN during training.

Fixes #20619

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21295.org.readthedocs.build/en/21295/

### CODE REVIEW

File: docs/source-pytorch/common/precision_basic.rst [hunk 1]
```diff
@@ -39,6 +39,14 @@ However, this setting can sometimes lead to unstable training.
 
     Trainer(precision="16-true")
 
+.. warning::
+
+    Float16 cannot represent values smaller than ~6e-5. Values like Adam's default ``eps=1e-8`` become zero, which can cause
+    NaN during training. Increase ``eps`` to 1e-4 or higher, and avoid extremely small values in your model weights and data.
```
Comment (deependujha) [inline docs/source-pytorch/common/precision_basic.rst:45]: when using `precision='16-true'`, should we also warn user when optimizer's eps can become zero and update it to `1e-4`?

Comment (bhimrazy) [reply docs/source-pytorch/common/precision_basic.rst:45]: Thanks for the suggestion @deependujha 
Currently, I’ve added these warnings considering both scenarios under the 16-bit Precision section.

Just to clarify — are you suggesting we should also add a programmatic runtime warning when eps might become zero?

Comment (deependujha) [reply docs/source-pytorch/common/precision_basic.rst:45]: yes, if you also think, it'll be a nice addition.

`Trainer` can check optimizer's eps and other attributes, and if using `16-true`, `warn` and update those attributes.

We can do it in another PR, if it makes sense.

**Review by Borda — APPROVED** [review]
